src/rgw/rgw_s3select.cc

```diff
@@ -624,39 +757,115 @@ int RGWSelectObj_ObjStore_S3::csv_processing(bufferlist& bl, off_t ofs, off_t le
 {
   int status = 0;
   
-  if (s->obj_size == 0) {
-    status = run_s3select(m_sql_query.c_str(), nullptr, 0);
+  if (s->obj_size == 0 || m_object_size_for_processing == 0) {
+    status = run_s3select_on_csv(m_sql_query.c_str(), nullptr, 0);
+    if(status<0){
+      return -EINVAL;
+    }
   } else {
     auto bl_len = bl.get_num_buffers();
     int i=0;
     for(auto& it : bl.buffers()) {
       ldpp_dout(this, 10) << "processing segment " << i << " out of " << bl_len << " off " << ofs
-                          << " len " << len << " obj-size " << s->obj_size << dendl;
-      if(it.length() == 0) {
+                          << " len " << len << " obj-size " << m_object_size_for_processing << dendl;
+      if(it.length() == 0 || len == 0){
         ldpp_dout(this, 10) << "s3select:it->_len is zero. segment " << i << " out of " << bl_len
-                            <<  " obj-size " << s->obj_size << dendl;
+                            <<  " obj-size " << m_object_size_for_processing << dendl;
         continue;
       }
+      //NOTE: the it.length() must be used (not len)
       m_aws_response_handler.update_processed_size(it.length());
-      status = run_s3select(m_sql_query.c_str(), &(it)[0], it.length());
+      status = run_s3select_on_csv(m_sql_query.c_str(), &(it)[0], it.length());
       if(status<0) {
+	return -EINVAL;
+      }
+      if(m_s3_csv_object.is_sql_limit_reached()) {
+	break;
+      }
+      i++;
+    }
+  }
+  if (m_aws_response_handler.get_processed_size() == uint64_t(m_object_size_for_processing) || m_s3_csv_object.is_sql_limit_reached()) {
+    if (status >=0) {
+      m_aws_response_handler.init_stats_response();
+      m_aws_response_handler.send_stats_response();
+      m_aws_response_handler.init_end_response();
+    }
+    if(m_s3_csv_object.is_sql_limit_reached()){
+    //TODO need to stop fetching chunks
+    ldpp_dout(this, 10) << "s3select : reached the limit :" << m_aws_response_handler.get_processed_size()  << dendl;
+    status = -2;
```

Comment:
please use posix error codes. in this case, `-ENOENT` doesn't seem like a good fit (same goes for `json_processing()` below)

nit: indentation